### ISSUE TITLE
Index Alma data triggered by webhook: daily updates and basedump

### ISSUE THREAD

**Issue by dr0i** [hbz/lobid-resources#1159]
Complements #1112 .
Fundamental to solve #1162.

@blackwinter has implemented a webhook which would call a (to be implemented) webhook listener to inform it about when the updates/basedump files are ready.

- [x] implement a webhook listener
- [x] index the data when webhook listener gets triggered
- [x] ~~put webhook listener into `monit` to ensure its living~~ (as it is part of lobid's web app it is already monitored)
- [x] get SSL certificate
- [x] restrict to SSL
- [x] configure webhook listener in the webhook

Deletions are atm not important but would complement this issue.

**Comment by blackwinter** [comment]
There are two possible webhooks (per type of run):

1. `baseline`: Gets triggered when a baseline run finishes.
2. `update`: Gets triggered when an update run finishes.

Both URLs may contain the following optional placeholders:

- `%<job>s`: The job to trigger (in our case: Jenkins job name); configurable per type of run (see above).
- `%<token>s`: A secret to protect the listener from unauthorized access (in our case: Jenkins Authentication Token); configurable per type of run (see above).
- `%<cause>s`: Some metadata pertaining to the run triggering the webhook; dynamic per run.

For Jenkins, our URL would look like:

`http://JENKINS_HOST/job/%<job>s/build?token=%<token>s&cause=%<cause>s`

**Comment by dr0i** [comment]
Webhook listener deployed, called like:
 `http://alma.lobid.org/resources/webhook/update-alma?token=$secret` and
`http://alma.lobid.org/resources/webhook/basedump-alma?token=$secret`.
Waiting for @blackwinter to configure the listener in their Webhook.

**Comment by dr0i** [comment]
Uh, wait: we should stick to SSL because of the secret token. Will assign @blackwinter again if the https URIs are properly configured.

**Comment by blackwinter** [comment]
The webhook endpoint should only enqueue an indexing job, not run it immediately (at least not synchronously). We're not going to wait for the indexing process to finish. If you can't/won't adjust your webhook listener, we'd need to change the trigger to avoid waiting for a response.

**Comment by dr0i** [comment]
Ah, yes I see, I did that wrong. Will fix this so that the endpoint is going to immediately ack when triggered.

**Comment by dr0i** [comment]
@blackwinter: Webhook Listener should work now as expected:
https://alma.lobid.org/resources/webhook/update-alma?token=$secret and
https://alma.lobid.org/resources/webhook/basedump-alma?token=$secret
Just try it. Note the restriction to use https.

**Comment by blackwinter** [comment]
Looks (almos) good here. Received `200 - OK: "... started ETL basedump"` when triggering `basedump-alma`; received `500 - Internal Server Error`, however, when triggering `update-alma` shortly afterwards.

**Comment by blackwinter** [comment]
Also verified that an incorrect `token` returns `403 - Forbidden`.

**Comment by dr0i** [comment]
re : 
> `500` ... when triggering update-alma shortly afterwards.

Hm, this should be a feature so that not too many requests are fired and no parallel ETL is ran, but I see your point: you think this should be handled by the listener  (do queuing etc), not the caller . Hm.

**Comment by blackwinter** [comment]
> you think this should be handled by the listener (do queuing etc), not the caller

Not necessarily, no. If you don't want/need a queue, it's fine. I was just surprised by the status code. Maybe something more specific might be more appropriate? (E.g., `423 Locked`, or `409 Conflict`, or just `503 Service Unavailable`.) Either way, it's entirely up to you; from a caller's standpoint, I don't _really_ care.

**Comment by blackwinter** [comment]
Just note that it might mask _actual_ server errors.

**Comment by dr0i** [comment]
Good point. Updated to respond with "423 Locked".

**Comment by blackwinter** [comment]
Sorry, can't confirm:

```none
2021-04-20T13:07:43.846 update: 200 - OK: "... started ETL update"
2021-04-20T13:07:47.662 update: 200 - OK: "... started ETL update"
2021-04-20T13:07:55.851 update: 200 - OK: "... started ETL update"
2021-04-20T13:08:24.342 baseline: 200 - OK: "... started ETL basedump"
2021-04-20T13:08:28.191 baseline: 500 - Internal Server Error
2021-04-20T13:08:32.765 update: 500 - Internal Server Error
2021-04-20T13:19:26.276 update: 500 - Internal Server Error
2021-04-20T13:19:30.701 update: 500 - Internal Server Error
```

1. Would have expected everything after line 1 to return `423` (unless the update was finished this fast!).
2. Would have expected all `500` to be `423`.

**Comment by dr0i** [comment]
Ups, hadn't deployed. Shouldn't do something quickly before lunch ....  Please try again.

**Comment by dr0i** [comment]
(the multiple started processes seem to be in connection with `update.xml.bgzf` beeing only 122 bytes small, resulting in just a little "locked" status. If you call the listener fast enough, it should work).

**Comment by dr0i** [comment]
Hm, something is strange and I cannot see the logs ETL program. Have to check.

**Comment by blackwinter** [comment]
Let me know if/when I shall give it another try. It's configured now and will be included when publishing starts again next week.

**Comment by dr0i** [comment]
Got logging to work (had to copy the log4j.xml to web/conf). Give it another try.

**Comment by blackwinter** [comment]
LGTM.

**Comment by dr0i** [comment]
Thx for reviewing and configuring! Closing.

**Comment by dr0i** [comment]
Uh. There was a bug in the creation of the index name, resulting in an overwriting of the existing production index. This happened 2 hours ago with the Webhook listener triggered from the BGFZ creation after NZ publishing.
Switched back to an index pre https://github.com/hbz/lobid-resources/pull/1214 (Thursday last week) , fixed the bug in 5caa61d and restarted deploying new index.

**Comment by dr0i** [comment]
ETLed the basedump of today and made it productive. Closing.
